File: java/437.path-sum-iii.java
import java.util.HashMap;
import java.util.Map;

/*
 * @lc app=leetcode id=437 lang=java
 *
 * [437] Path Sum III
 */

// @lc code=start
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode() {} TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) { this.val = val; this.left
 * = left; this.right = right; } }
 */
class Solution {
    Map<Integer, Integer> past_sum = new HashMap<>() {
        {
            put(0, 1);
        }
    };
    int res = 0;

    public int pathSum(TreeNode root, int targetSum) {
        preorderTras(root, targetSum, 0);
        return res;
    }

    public void preorderTras(TreeNode root, final int targetSum, int current_sum) {
        if (root == null) {
            return;
        }
        current_sum += root.val;
        if (past_sum.containsKey(current_sum - targetSum)) {
            res += past_sum.get(current_sum - targetSum);
        }
        past_sum.put(current_sum, past_sum.getOrDefault(current_sum, 0) + 1);
        preorderTras(root.left, targetSum, current_sum);
        preorderTras(root.right, targetSum, current_sum);
        past_sum.put(current_sum, past_sum.get(current_sum) - 1);
    }
}
// @lc code=end

File: cpp/437.path-sum-iii.cpp
/*
 * @lc app=leetcode id=437 lang=cpp
 *
 * [437] Path Sum III
 */
#include <unordered_map>
#include <vector>
using namespace std;
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
    unordered_map<int, int> past_sum = {{0, 1}};

public:
    int pathSum(TreeNode *root, int targetSum)
    {
        int res = 0;
        preorderTras(res, root, targetSum, 0);
        return res;
    }
    void preorderTras(int &res, TreeNode *root, const int &targetSum, int current_sum)
    {
        if (!root)
        {
            return;
        }
        current_sum += root->val;
        if (past_sum.count(current_sum - targetSum))
        {
            res += past_sum[current_sum - targetSum];
        }
        ++past_sum[current_sum];
        preorderTras(res, root->left, targetSum, current_sum);
        preorderTras(res, root->right, targetSum, current_sum);
        --past_sum[current_sum];
    }
};
// @lc code=end
